### ISSUE TITLE
Support Typescript incremental build

### ISSUE THREAD

**Issue by glebmachine** [angular/angular-cli#13941]
# 🚀 Feature request

### Command (mark with an `x`)

```
- [ ] new
- [x] build
- [x] serve
- [x] test
- [x] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [ ] xi18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc
```

### Description
Support of Typescript incremental build https://devblogs.microsoft.com/typescript/announcing-typescript-3-4-rc/

**Comment by jimbuck** [comment]
[3.4 is officially out.](https://devblogs.microsoft.com/typescript/announcing-typescript-3-4/) Out of curiosity may I ask what is currently causing this to be marked as "blocked"?

**Comment by clydin** [comment]
Angular itself needs to support a TS version before it can be integrated into the tooling.

**Comment by RicardoVaranda** [comment]
@alan-agius4 @filipesilva Hi Guys! Congrats on the new release! Just wondering if there has been any development with this improvement since TS 3.4 is now supported.

**Comment by MartinJohns** [comment]
> since TS 3.4 is now supported.

@RicardoVaranda Angular 8.0.0 supports TypeScript 3.3, but not 3.4 (which introduced the incremental flag).

https://github.com/angular/angular/blob/master/CHANGELOG.md

> add support for TypeScript 3.3

**Comment by tomardern** [comment]
@MartinJohns Interestingly, 3.4 looks like it was merged into angular-8-beta-12: https://github.com/angular/angular/commit/ef853367194c77f505a0705e45f5670725873da1

**Comment by RicardoVaranda** [comment]
Can confirm this, upon doing ng update it actually installed `3.4.2`

**Comment by MarcosMeli** [comment]
Is important to note that 3.4 introduce a regression that make type checking in compilation pain slow in some scenarios. 
Now 3.5 is out with this regression fixed: https://devblogs.microsoft.com/typescript/announcing-typescript-3-5/ 

> **Speed improvements**
> TypeScript 3.5 introduces several optimizations around type-checking and incremental builds.
> **Type-checking speed-ups**
> Much of the expressivity of our type system comes with a cost – any more work that we expect the compiler to do translates to longer compile times. Unfortunately, as part of a bug fix in TypeScript 3.4 we accidentally introduced a regression that could lead to an explosion in how much work the type-checker did, and in turn, type-checking time. The most-impacted set of users were those using the styled-components library. This regression was serious not just because it led to much higher build times for TypeScript code, but because editor operations for both TypeScript and JavaScript users became unbearably slow.

**Comment by filipesilva** [comment]
Angular version 8 supports (and requires) TypeScript 3.4.x. I haven't look at this again, but at the time the only problem I saw was https://github.com/angular/angular/pull/29872.

We have to try and figure out what else (if anything) is preventing the incremental cache from being used and written.

I'm not sure it can be used in the AOT compiler though, so keep that in mind. Maybe it works, maybe it doesn't. 

If anyone is interested in having a look at this, here's a good way to get started:
```
git clone https://github.com/angular/angular-cli
cd angular-cli
yarn
```

Then open `packages/angular_devkit/build_angular/test/browser/works_spec_large.ts` and replace `it(` with `fit(`. Also remove the `afterEach`. Now whenever you run `yarn test-large`, only this test runs. The results for this test will be in a temporary folder in `tests/angular_devkit/build_angular`.

At this point you'd just have to hunker down and go figure out what the logic for incremental builds in TS is trying to do, and how to work with it. There's not a lot of docs around it. When I was looking into this, I just added a lot of logs to the TS source in `node_modules/typescript`.

We use a in-memory file system for TS in `packages/ngtools/webpack`. So even if the cache file is being written, if TS is writing it then it will stay in memory. It would have to be written there somewhere. Check the `packages/ngtools/webpack/src/compiler_host.ts` for the FS abstraction used by TS. Maybe log some stuff in the writeFile method.

**Comment by FDIM** [comment]
From what I understand for incremental build to work and actually save time, files have to be emitted to outDir. If the incremental flag is true, there will be an extra tsconfig.tsbuildinfo file that holds hashes of involved files.

When we build with the CLI, outDir does not seem to be used. @filipesilva is this part also using in-memory file system during build? If it is virtualized, do you know if differential build reuses the same virtual FS or a new is created? 

I am quite curious about this, for my little chrome extension (for e2e tests), if I run tsc command and pass project tsconfig, first build is ~15s, 2nd ~5s

**Comment by filipesilva** [comment]
@FDIM I think realistically what would happen is that we'd have to allow that file to be loaded between builds. 

Maybe to make sure it's loaded it is enough to give it a custom path instead of the outDir default. But at the end of each build it would still need to be persisted to disk with custom logic.

I am not sure how much this would help for differential loading as each build does use different ES targets. At that point it depends on how much TS itself caches for those cases.

We don't use TS loader, mind you. We have our own loader tailored to the Angular AOT compiler. So I don't think ts-loader supporting it blocks us.

**Comment by filipesilva** [comment]
For reference, the incremental API issue on TS is https://github.com/microsoft/TypeScript/issues/29978.

**Comment by FDIM** [comment]
My bad regarding ts-loader, I should have realized that when tinkering with the loader in CLI.

I am fairly sure that writing tsconfig.tsbuildinfo file to disk would not be enough, it also needs already built files from last time to avoid compilation. Anyway, I did manage to output build files to disk after running ng build, but tsbuildinfo was not part of the output - not sure why.

**Comment by kukjevov** [comment]
Typescript already finished their implementation, so it should be possible to try it here :)

**Comment by pppdns** [comment]
Seems like there are no more roadblocks for Angular to support incremental builds.

Some related changes that happened since then:

Provide --build/--incremental APIs for build tools:
https://github.com/microsoft/TypeScript/issues/29978

Api for tsc --build and --incremental:
https://github.com/microsoft/TypeScript/pull/31432

Docs are in progress:
https://github.com/microsoft/TypeScript/issues/31849

TS 3.5 Release notes about speed improvements and fixed regression from 3.4:
https://devblogs.microsoft.com/typescript/announcing-typescript-3-5-rc/#speed-improvements

**Comment by wottpal** [comment]
> Seems like there are no more roadblocks for Angular to support incremental builds.

@pppdns Does this mean I just have to set `"compilerOptions": { "incremental": true }` in my `tsconfig.json` and it works or is there still something to do on the Angular-side?

**Comment by snsakib** [comment]
@wottpal 
hope it's simple as that

**Comment by FDIM** [comment]
Unfortunately to improve cold start/initial compile, additional work is needed.

Api wise, new cache dir option should be enough, but it still need to be implemented.

**Comment by jukibom** [comment]
@FDIM I may be reading this wrong, but doesn't [this](https://github.com/microsoft/TypeScript/pull/31432/files#diff-af8fc5dff8fdc68132f4be74692af69cR4583) permit arbitrary compiler options?

If so, doesn't `tsBuildInfoFile` provide the ability to specify the location of build cache?

**Comment by kukjevov** [comment]
Anything new here ? Typescript 3.5 is now supported by Angular

**Comment by moohkooh** [comment]
would be nice to have support of incremental build for ngc

**Comment by kukjevov** [comment]
Also with incremental build it would be not nice but super duper to have support of composite projects. I think it would heavily help large projects to split their projects into smaller subprojects. As far as i know typescript created API for both incremental and "composite" build mode in 3.5.

Well i think it has to supported in compiler-cli first so maybe this issue should go to angular project?
:)

**Comment by aruballo** [comment]
Any updates on this? Looking to implement this in a complex project that I recently upgraded to Angular 8.2. I added the incremental and tsBuildinfoFile properties but I don't see a difference (I dont even see any new files generated anywhere)

**Comment by FDIM** [comment]
I am not sure what happened, but the API for incremental option/program is not available in TS 3.5, only in 3.6. It will soon be released though (23rd of August according to road map).

So for now I'd say this is blocked until angular packages depend on 3.6. Afterwards 2 solutions will be needed: 
- find a way to configure current build to emit compiled ts files and tsbuildinfo in cache/out dir (side note: Maybe ivy could use this to cache factories as well to further improve cold start time)
- find a way to create incremental program that uses existing tsbuildinfo and feed that to current build. As far as I understand this program could be used as oldProgram (not sure though...) 

As for cache dir @filipesilva , I guess it would be best to reuse `outDir` from ts config instead of adding new to angular.json? Currently angular compiler plugin is resetting this value and if it is set, build fails with a bit unclear message.

**Comment by aruballo** [comment]
@FDIM Where did you hear that? The documentation for typescript seem to indicate this is available in 3.4:

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html

**Comment by clydin** [comment]
The functionality was available via `tsc` (the TypeScript command line interface) in 3.4.  What is needed here for it to be considered for inclusion is programmatic access.

**Comment by kukjevov** [comment]
@FDIM yes it is interesting about that API and version of typescript, because i followed release of 3.5 and fix was not present at time of release 3.5, but still they put it in changelog of 3.5 even thou i read some guy complaint that there is no API for incremental build available in 3.5, so it looks like some mess in release log, but it will be in 3.6 definitely :)

**Comment by FDIM** [comment]
@aruballo I've tried to use the api to see how it works, only to realize that it is not there

**Comment by aruballo** [comment]
How weird, I guess I'll have to wait for 3.6 before using this then. Thanks @FDIM

**Comment by kukjevov** [comment]
https://devblogs.microsoft.com/typescript/announcing-typescript-3-6-rc/

Well it looks like it is finally coming for real to typescript, i would love to see it in angular compiler, both `--build` and `--incremental` support :)

**Comment by chadbr** [comment]
https://devblogs.microsoft.com/typescript/announcing-typescript-3-6/

🔥🔥🔥

**Comment by kukjevov** [comment]
Any movement here? Or probably this is waiting for release of 9th version and there is primary focus for that now, probably. :)

**Comment by filipesilva** [comment]
I should set expectations regarding the potential usefulness of supporting TS's `--incremental` functionality right now. As a recap, this functionality from TS allows a typescript compilation to reuse artifacts from a previous compilation in order to be faster. This doesn't apply to watch mode. It only applies to whole new compilations (e.g. calling the command from the console again).

In an Angular CLI context this would mean that some calls to `ng build`, `ng serve` and `ng test` might be faster after the first call. Ideally you'd want your slowest builds, usually the production builds, to get faster by using this new typescript functionality. But there are a lot of caveats here. 

First and foremost is that the typescript compilation is only one part of building an app, and not necessarily what takes longer. Then there's the fact that AOT compilation is not just a Typescript compilation, there's many other things happening there that have nothing to do with compiling TS. Another thing to take into consideration is that CI builds wouldn't use this either unless you were very careful in managing the TS cache.

Realistically, this means that production builds wouldn't see any improvement because they use AOT and a big chunk of the time is spent in bundle optimization. JIT development builds are what would benefit from TS's `--incremental` functionality. But with Ivy it's already preferable to develop in AOT, so no benefit again.

I feel this feature looks very desirable because it sounds like it would make all builds much faster. But the reality is that it would make some of the faster builds a bit faster, while the others would remain the same. Notably rebuilds on watch mode are entirely unaffected, because this functionality isn't related to watch mode.

**Comment by kukjevov** [comment]
Thanks for clarification. What about composite projects ? For me this would be more beneficial than incremental, because this allows split big projects into "smaller" parts.

BTW ts-loader managed to use incremental api somehow https://github.com/TypeStrong/ts-loader/issues/913, but i do understand that one thing is angular compilation and different thing is typescript compilation.

**Comment by filipesilva** [comment]
@kukjevov we don't have any support for composite projects at the moment so I don't think that would be a factor here. I'm not sure how it would work in the app case anyway... within the app rebuilds are already incremental, and outside the apps there are only libs, and libs need their own build process that also involves AOT.

AOT means "Ahead of Time" and is used in contrast with JIT, which means "Just in Time". Template compilation happens in the browser when you are not doing AOT compilation (e.g. you are in JIT). This means your app is loaded on the browser, then it reads all the templates (html and css), and then converts them into JS code to be used by Angular. 

This increases the app startup time and also needs the compiler needs to be bundled with your application. When you do AOT compilation, you compile templates when you build your app locally, and then you don't have to bundle the compiler at all or spend time in the browser compiling things. 

The difference between AOT compilation and TS compilation is that AOT compilation also compiles your templates in addition to doing TS compilation. The AOT compiler is usually referred to as `ngc` because that's the binary that `@angular/compiler-cli` exposes for AOT compilation.

**Comment by kukjevov** [comment]
Hi @filipesilva, thx for your answer :).

I know how is AOT working and what it means :). Well i would like to explain you my case.

I have one bigger project. It has parts that are not changed often and are not dependent on many other files. These files, or better said this directory is ideal canditate for *reference* project. Because it has not dependencies outside of this directory and is not often changed. So if you use composite project, it will store build output information and reuse them for next compilation. Which is great because it saves resources. I have few directories that could be turned into this type of *reference* project. This is project which is using JIT, so it is compiled only using plain Typescript. So far i have been using *awesome typescript loader*, but i would like to try *ts-loader* which looks like already have support for incremental and composite projects (ts-loader team somehow managed to implement also watch api with composite and incremental api).

I plan to migrate it to angular 9 with ivy as soon as possible :). I believe it will be faster, because i already have *small* projects running angular 9.0.0-next.9 with aot and ivy and it is total brutal fast, but project is smaller. So maybe this will not be necessary if i migrate to ivy and 9, but still it would be awesome to have some option to split projects into smaller parts and dependent parts would reuse *metainfo* or something from last compilation so typechecking, template compilation and other stuff does not have to be done everytime, but maybe this is solved in ivy already, because as i said those smaller projects are insanly fast.

Still everything is awesome i just asking if something like that is planned :).

**Comment by alxhub** [comment]
@filipesilva is correct that `composite` projects are not supported currently. There is no strong technical reason for this, it's just not been a priority since there are other ways Angular users split up compilations (e.g. `nx` workspaces, Bazel) and we need to ensure the compiler does a good job for these circumstances first. Once the general multi-project story is in a good place, it will make sense to implement the TS layer.

The same is true for the incremental build cache. One of the big requirements is having strong incremental rebuild semantics under `--watch`. We have a roadmap for this that we're part of the way through implementing.

It's *very* important that we get the semantics right when we move to an incremental cache. Right now under `--watch` we have a safety net: if the compiler has a bug and does not handle an incremental change correctly, the user will just kill the `--watch` compilation (e.g. by killing `ng serve`) and rerun it. This safety net doesn't exist with an incremental build cache - if we process a change incorrectly, the mistake will persist until you turn off incremental compilation. Therefore I want to move slowly and deliberately when implementing this feature. I'd rather err on the side of not supporting it until we're absolutely sure our semantics are perfect.

**Comment by jukibom** [comment]
Just dropping in a use-case here for composite project support. 

We have a typescript monorepo of which one of the packages is the main front-end angular app and which consumes a number of other application-like packages in a yarn workspace (and those packages have other dependencies in the repo, such as a core ECS package, a general service interaction layer etc). Building any one of these applications will auto-build any composite dependencies it needs (and their dependencies and so on), so conceptually we only need to run a build on any one top-level application a developer's working on.

For the angular frontend we have to first build all packages it depends on and get a build watcher going on each application a developer is working on. These are better than nothing but pretty slow and sometimes unreliable. It's also just generally a bit of a pain for devs to get an environment set up. So in our case it's not so much the incremental build cache (though that would be nice!) but the ability to allow typescript to recognise other composite projects in the repo with the `build` flag that we're really interested in.

**Comment by andreialecu** [comment]
I have a different use case regarding composite projects where I only have non angular related code in separate projects under a yarn workspace.

I was expecting to be able to use `@ngtools/webpack` with a tsconfig that defines the reference and enables composite mode, but it looks like it doesn't work properly, and doesn't build referenced projects. The separate projects output their compiled js to directories such as `/lib` underneath them, and everything is resolved through package.json references.

As a workaround, I need to build the referenced projects manually before starting the main angular webpack build. 

I'm wondering if `@ngtools/webpack` could be updated to support this minimal scenario of composite projects until full AOT support is added.

**Comment by glebmachine** [comment]
Any updates on this?

**Comment by alan-agius4** [comment]
Transfering to the FW repo as the compiler doesn't support TypeScript incremental API. 

Example it uses `ts.createProgram`, instead of `ts.createIncrementalProgram` with the latter is require to use the tsbuild info.

I did experiment a bit using the NGTSC plugin API with allows for a program to be supplied and patching on a small project https://github.com/ngrx/platform/tree/master/modules/component/src I saw the compilation to go from `5s` to `2.5s`.
